src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp

```diff
@@ -534,13 +534,31 @@ void os::verify_stack_alignment() {
 }
 #endif
 
+static THREAD_LOCAL WXMode _wx_state = WXUnknown;
 int os::extra_bang_size_in_bytes() {
   // AArch64 does not require the additional stack bang.
   return 0;
 }
 
-void os::current_thread_enable_wx(WXMode mode) {
-  pthread_jit_write_protect_np(mode == WXExec);
+WXMode os::current_thread_change_wx(WXMode new_state) {
+  if (_wx_state == WXUnknown) {
+    pthread_jit_write_protect_np(false /* not executable but writable */);
+    _wx_state = WXWrite;
+  }
+  WXMode old = _wx_state;
+  if (_wx_state != new_state) {
+    _wx_state = new_state;
+    pthread_jit_write_protect_np(_wx_state == WXExec);
+  }
+  return old;
+}
+
+void os::current_thread_reset_wx() {
+  current_thread_change_wx(WXWrite);
+}
```

Comment:
I find the naming a  bit misleading. You use this as initialization, so I would call it "init" something.

Then, I'm not sure it is even needed. I know you just transformed it from the original `init_wx()`, so the question is directed more at the original authors (@AntonKozlov?).

AFAIU we use this to initialize wxstate for newly attached threads to "dont execute". But should this not already be the case? And if its not - e.g. because that thread had been calling into another library that also does call generated code - is it not impolite to switch the state to "executable false"? I know this is highly unlikely, I just try to understand.

Reply:
> I find the naming a bit misleading. You use this as initialization, so I would call it "init" something.
You're correct.

> AFAIU we use this to initialize wxstate for newly attached threads to "dont execute". But should this not already be the case? 

No, it does not seem to be the case: https://github.com/dotnet/runtime/issues/41991  (and the man page does not give a default value either)

> And if its not - e.g. because that thread had been calling into another library that also does call generated code - is it not impolite to switch the state to "executable false"? I know this is highly unlikely, I just try to understand.

This might be an issue when calling another library with a JIT. But this seems to be highly unlikely. Especially as this would have already been an issue.